### Import Library

In [1]:
from sklearn.linear_model import LinearRegression,Ridge,Lasso
from sklearn.metrics import r2_score,mean_absolute_error,mean_squared_error
from sklearn.model_selection import RandomizedSearchCV
from catboost import CatBoostRegressor
from xgboost import XGBRFRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import KneighborsRegressor

import warnings
import pandas as pd
import numpy as np
import matplotlib as plt 
import seaborn as sb

### Import csv file

In [26]:
df = pd.read_csv('E:\\Neoron\\Programming_Practice\\Machine_Learning_Project\\ml_Score_Prediction\\Data\\stud.csv')
df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

### Preparing X and Y 


In [27]:
x = df.drop(['math_score'],axis=1)

In [28]:
y = df['math_score']

In [29]:
# create column transformation 
num_feature = x.select_dtypes(exclude=object).columns
cat_feature = x.select_dtypes(include=object).columns

from sklearn.preprocessing import OneHotEncoder , StandardScaler
from sklearn.compose import ColumnTransformer

numeric_transformer = StandardScaler()
oh_transformer = OneHotEncoder()

preprocessor = ColumnTransformer(
    [
        ("OneHotEncoder",oh_transformer,cat_feature),
        ("StandardScaler",numeric_transformer,num_feature),
    ]
)

In [30]:
x = preprocessor.fit_transform(x)

In [33]:
# Separate dataset into train and test 
from sklearn.model_selection import train_test_split

x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2,random_state=0)
x_train.shape ,x_test.shape

((800, 19), (200, 19))

### Create an evaluate function to give all metrics after training

In [46]:
'''def evaluate_model(y_true,y_pred):
    r2 = r2_score(y_true,y_pred)
    mae = mean_absolute_error(y_true,y_pred)
    mse = mean_squared_error(y_true,y_pred)
    rmse = np.sqrt(mse)
    print(f"R2 Score : {r2}")
    print(f"Mean Absolute Error : {mae}")
    print(f"Mean Squared Error : {mse}")
    print(f"Mean Root Squared Error : {rmse}")'''
    
    
def evaluate_model(y_true,y_pred):
        r2 = r2_score(y_true,y_pred)
        mae = mean_absolute_error(y_true,y_pred)
        mse = mean_squared_error(y_true,y_pred)
        rmse = np.sqrt(mse)
        return r2, mae, mse, rmse

In [47]:
from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor
from xgboost import XGBRegressor
from sklearn. ensemble import AdaBoostRegressor
from sklearn.ensemble import RandomForestRegressor

In [48]:
models = {
    "Linear Regression": LinearRegression(),
    "Lasso": Lasso(),
    "Ridge": Ridge(),
    "K-Neighbors Regressor": KNeighborsRegressor(),
    "Decision Tree": DecisionTreeRegressor(),
    "Random Forest Regressor": RandomForestRegressor(),
    "XGBRegressor": XGBRegressor(), 
    "CatBoosting Regressor": CatBoostRegressor(verbose=False),
    "AdaBoost Regressor": AdaBoostRegressor()
}
model_list = []
r2_list =[]

for i in range(len(list(models))):
    model = list(models.values())[i]
    model.fit(x_train, y_train) # Train model

    # Make predictions
    y_train_pred = model.predict(x_train)
    y_test_pred = model.predict(x_test)
    
    # Evaluate Train and Test dataset
    r2 , mae, mse ,rmse= evaluate_model(y_train, y_train_pred)

    model_test_r2 , model_test_mae, model_test_mse,model_test_rmse = evaluate_model(y_test, y_test_pred)

    
    print(list(models.keys())[i])
    model_list.append(list(models.keys())[i])
    
    print('Model performance for Training set')
    print("- Root Mean Squared Error: {:.4f}".format(rmse))
    print("- R2 Score: {:.4f}".format(r2))
    print("- Mean Absolute Error: {:.4f}".format(mae))
    print("- Mean Squared Error: {:.4f}".format(mse))

    print('----------------------------------')
    
    print('Model performance for Test set')
    print("- Root Mean Squared Error: {:.4f}".format(model_test_rmse))
    print("- Mean Absolute Error: {:.4f}".format(model_test_mae))
    print("- R2 Score: {:.4f}".format(model_test_r2))
    print("- Mean Squared Error: {:.4f}".format(model_test_mse))
    r2_list.append(model_test_r2)
    
    print('='*35)


Linear Regression
Model performance for Training set
- Root Mean Squared Error: 5.2899
- R2 Score: 0.8786
- Mean Absolute Error: 4.2364
- Mean Squared Error: 27.9833
----------------------------------
Model performance for Test set
- Root Mean Squared Error: 5.5854
- Mean Absolute Error: 4.3975
- R2 Score: 0.8623
- Mean Squared Error: 31.1970
Lasso
Model performance for Training set
- Root Mean Squared Error: 6.6102
- R2 Score: 0.8104
- Mean Absolute Error: 5.1990
- Mean Squared Error: 43.6951
----------------------------------
Model performance for Test set
- Root Mean Squared Error: 6.7021
- Mean Absolute Error: 5.3566
- R2 Score: 0.8018
- Mean Squared Error: 44.9186
Ridge
Model performance for Training set
- Root Mean Squared Error: 5.2818
- R2 Score: 0.8789
- Mean Absolute Error: 4.2268
- Mean Squared Error: 27.8974
----------------------------------
Model performance for Test set
- Root Mean Squared Error: 5.5674
- Mean Absolute Error: 4.3880
- R2 Score: 0.8632
- Mean Squared Erro

### Conclusions:




1. Root Mean Squared Error (RMSE): The RMSE measures the average magnitude of errors (the differences between predicted values and actual values) in your model's predictions. In this case, the RMSE is approximately 5.2899. A lower RMSE indicates better model performance, as it signifies that the model's predictions are closer to the actual values on average.

2. R-squared (R2) Score: The R2 score, also known as the coefficient of determination, quantifies how well the independent variables in your model explain the variance in the dependent variable. An R2 score of 0.8786 means that approximately 87.86% of the variance in the dependent variable is explained by the independent variables in your model. A higher R2 score indicates a better fit of the model to the data.

3. Mean Absolute Error (MAE): The MAE represents the average absolute differences between the predicted values and the actual values. In this case, the MAE is approximately 4.2364. Like RMSE, a lower MAE indicates better model performance, with smaller errors between predictions and actual values.

4. Mean Squared Error (MSE): The MSE is the average of the squared differences between predicted and actual values. It is often used as a loss function during model training. The MSE in your case is approximately 27.9833. Similar to RMSE and MAE, a lower MSE indicates better model performance, with smaller squared errors between predictions and actual values.

In summary, based on these metrics:

- Your model has a relatively low RMSE, indicating that, on average, its predictions are close to the actual values.
- The R2 score of 0.8786 suggests that the model explains a significant portion of the variance in the dependent variable.
- The MAE and MSE values are also relatively low, further indicating good model performance in terms of prediction accuracy.

However, it's important to consider the specific problem and context when evaluating these metrics, as what constitutes "good" performance can vary depending on the application. Additionally, it's a good practice to compare these metrics to a baseline or other models to assess whether your model's performance is satisfactory.

In [49]:
# Result of model
result = pd.DataFrame({'Model':model_list,'R2_score':r2_list})
result = result.sort_values(by='R2_score',ascending=False)
result

Model  R2_score
2                    Ridge  0.863213
0        Linear Regression  0.862327
7    CatBoosting Regressor  0.857797
5  Random Forest Regressor  0.856879
8       AdaBoost Regressor  0.822903
6             XGBRegressor  0.818506
1                    Lasso  0.801773
3    K-Neighbors Regressor  0.790108
4            Decision Tree  0.731357